engine/src/main/java/org/camunda/bpm/application/impl/JakartaServletProcessApplicationDeployer.java

```diff
@@ -0,0 +1,90 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.application.impl;
+
+import jakarta.servlet.ServletContainerInitializer;
+import jakarta.servlet.ServletContext;
+import jakarta.servlet.ServletContextListener;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.annotation.HandlesTypes;
+import java.util.HashSet;
+import java.util.Set;
+import org.camunda.bpm.application.AbstractProcessApplication;
+import org.camunda.bpm.application.ProcessApplication;
+import org.camunda.bpm.engine.impl.ProcessEngineLogger;
+
+/**
+ * <p>This class is an implementation of {@link ServletContainerInitializer} and
+ * is notified whenever a subclass of {@link ServletProcessApplication} annotated
+ * with the {@link ProcessApplication} annotation is deployed. In such an event,
+ * we automatically add the class as {@link ServletContextListener} to the
+ * {@link ServletContext}.</p>
+ *
+ * <p><strong>NOTE:</strong> Only works with Servlet 5.0 or better.</p>
+ *
+ */
+@HandlesTypes(ProcessApplication.class)
+public class JakartaServletProcessApplicationDeployer implements ServletContainerInitializer {
+
+  private static ProcessApplicationLogger LOG = ProcessEngineLogger.PROCESS_APPLICATION_LOGGER;
+
+  @Override
+  public void onStartup(Set<Class<?>> c, ServletContext ctx) throws ServletException {
+    if(c == null || c.isEmpty()) {
+      // skip deployments that do not carry a PA
+      return;
+
+    }
+
+    if (c.contains(ProcessApplication.class)) {
+      // this is a workaround for a bug in WebSphere-8.5 who
+      // ships the annotation itself as part of the discovered classes.
+
+      // copy into a fresh Set as we don't know if the original Set is mutable or immutable.
+      c = new HashSet<Class<?>>(c);
+
+      // and now remove the annotation itself.
+      c.remove(ProcessApplication.class);
+    }
+
+
+    String contextPath = ctx.getContextPath();
+    if(c.size() > 1) {
+      // a deployment must only contain a single PA
+      throw new ServletException(LOG.multiplePasException(c, contextPath));
+
+    } else if(c.size() == 1) {
+      Class<?> paClass = c.iterator().next();
+
+      // validate whether it is a legal Process Application
+      if(!AbstractProcessApplication.class.isAssignableFrom(paClass)) {
+        throw new ServletException(LOG.paWrongTypeException(paClass));
+      }
+
+      // add it as listener if it's a ServletProcessApplication
+      if(ServletProcessApplication.class.isAssignableFrom(paClass)) {
+        LOG.detectedPa(paClass);
+        ctx.addListener(paClass.getName());
+      }
+    }
+    else {
+      LOG.servletDeployerNoPaFound(contextPath);
+    }
+
+  }
+
+}
```

Comment:
```suggestion
@HandlesTypes(ProcessApplication.class)
public class JakartaServletProcessApplicationDeployer extends AbstractServletProcessApplicationDeployer implements ServletContainerInitializer {

  private static ProcessApplicationLogger LOG = ProcessEngineLogger.PROCESS_APPLICATION_LOGGER;

  @Override
  public void onStartup(Set<Class<?>> c, ServletContext ctx) throws ServletException {
    try {
      super.onStartup(c, ctx.getContextPath(), ctx::addListener);
    } catch (Exception e) {
      if (e instanceof  ServletException) {
        throw (ServletException) e;
      } // we don't care about other exceptions
    }
  }
  @Override
  protected ServletException throwMultipleServletException(Set<Class<?>> c, String contextPath) {
    return new ServletException(LOG.multiplePasException(c, contextPath));
  }

  @Override
  protected ServletException throwPaWrongTypeException(Class<?> paClass) {
    return new ServletException(LOG.paWrongTypeException(paClass));
  }

}
```
This class duplicates code which is not ideal (we might miss to change the code at both places when touching this again). Can we refactor it even more? Something along the following lines ⬆️
Also the workaround for WAS 8.5 is not needed anymore, right?

This is just an example. There might be other places as well where we can reduce code duplications.

Reply:
Good call, definitely worth a closer look 👍
I can have another look at the other Jakarta equivalents to extract as much code as possible.

Reply:
I adjusted the servlet deployer and also some other classes to extract more business logic into common super classes 👍